harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+        self._start_ts = time.time()
+        self._start_vals = psutil.disk_io_counters()
+
+        # Initialize accessible disk paths.
+        self._paths = []
+        for path in self._disk_paths:
+            try:
+                psutil.disk_usage(path)
+                self._paths.append(path)
+            except Exception:
+                pass
+
+        super().__init__()
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        ts = time.time()
+        vals = psutil.disk_io_counters()
+
+        read_thru = (vals.read_bytes - self._start_vals.read_bytes) / (ts - self._start_ts)
+        write_thru = (vals.write_bytes - self._start_vals.write_bytes) / (ts - self._start_ts)
+        iops = (vals.read_count + vals.write_count) - (
+            self._start_vals.read_count + self._start_vals.write_count
+        ) / (ts - self._start_ts)
+        self._start_ts, self._start_vals = ts, vals
+
+        metrics = {
+            "disk_iops": iops,
+            "disk_throughput_read": read_thru,
+            "disk_throughput_write": write_thru,
+        }
+
+        for path in self._paths:
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        return metrics
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        free_mem_bytes = psutil.virtual_memory().available
+        return {
+            "free_memory": free_mem_bytes / 1e9,
+        }
+
+
+class _CPU(_MetricGroupCollector):
+    group = "cpu"
+
+    def collect(self) -> typing.Dict[str, typing.Any]:
+        cpu_util = psutil.cpu_percent()
+        return {
+            "cpu_util_simple": cpu_util,
+        }
+
+
+class _GPU(_MetricGroupCollector):
+    group = "gpu"
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self._pynvml_device_handles: typing.Dict[str, typing.Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning("pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        """
+        Call the NVML library methods we'll be using to validate that they are accessible.
+        Sometimes NVML initializes successfully but individual device methods will fail.
+        """
+        try:
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                uuid = pynvml.nvmlDeviceGetUUID(handle)
+                pynvml.nvmlDeviceGetMemoryInfo(handle)
+                pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml_device_handles[uuid] = handle
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
```

Comment:
I'm always uncomfortable with bare exceptions (they're https://google.github.io/styleguide/pyguide.html#244-decision for a reason). I'd prefer to start with a minimal set of exceptions and expand them as we discover them.

Besides the `pynvml.NVMLError`, what are you wanting to catch?

Reply:
removed this, just catching the only `NVMLError` now